### PR TITLE
[Security] Update password upgrader listener to work with the new UserBadge

### PR DESCRIPTION
**Pull request by wouterj**
| Q             | A
| ------------- | ---
| Branch?       | 5.2
| Bug fix?      | yes
| New feature?  | yes
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        | -

While working on a new amazing `make:auth` maker, @jrushlow discovered that we forgot to update the `PasswordUpgradeBadge` with the 5.2 `UserBadge` changes (ref https://github.com/symfony/symfony/pull/37846).

This PR fixes it, by making the password upgrader optional and falling back to the user provider instead. Without these changes, each authenticator still needs to know the user repository/user provider in order to pass it to `PasswordUpgradeBadge`.

I'm sorry for catching this soo late in the release cycle. There is a BC break involved here, but it's (a) very unlikely to impact application code and (b) in an experimental class.

### CODE REVIEW

File: src/Symfony/Component/Security/CHANGELOG.md [hunk 1]
```diff
@@ -15,6 +15,7 @@ CHANGELOG
  * Added `LoginThrottlingListener`.
  * Added `LoginLinkAuthenticator`.
  * Moved methods `supports()` and `authenticate()` from `AbstractListener` to `FirewallListenerInterface`.
+ * [BC break] `$passwordUpgrader` in `PasswordUpgradeBadge` is now nullable (in 2nd argument and `getPasswordUpgrader()`)
```
Comment (chalasr) [inline src/Symfony/Component/Security/CHANGELOG.md]: I don't think it's a BC break actually as that constructor arg still accepts the same type, so we just allow omitting it or using null.

**Review by chalasr — APPROVED** [review]
Nice catch!

File: src/Symfony/Component/Security/CHANGELOG.md (hunk 1 above)
Comment (derrabus) [reply src/Symfony/Component/Security/CHANGELOG.md]: Yes, but the fact that `getPasswordUpgrader()` might return `null` now actually is a small BC break because code calling that method might not expect that.

File: src/Symfony/Component/Security/Http/Tests/EventListener/PasswordMigratingListenerTest.php [hunk 2]
```diff
@@ -91,3 +103,22 @@ private function createEvent(PassportInterface $passport)
...
+    }
+
+    public function refreshUser(UserInterface $user)
+    {
+    }
+
+    public function supportsClass(string $class)
+    {
+    }
+}
```
Comment (derrabus) [inline src/Symfony/Component/Security/Http/Tests/EventListener/PasswordMigratingListenerTest.php]:
```suggestion
abstract class MigratingUserProvider implements UserProviderInterface, PasswordUpgraderInterface
{
}
```

Because you only need this class for mocking, you can just set it abstract and remove the empty method bodies.

File: src/Symfony/Component/Security/CHANGELOG.md (hunk 1 above)
Comment (chalasr) [reply src/Symfony/Component/Security/CHANGELOG.md]: Indeed I missed the `getPasswordUpgrader()` part, I'd reword the message to talk only about it then.

File: src/Symfony/Component/Security/Http/Tests/EventListener/PasswordMigratingListenerTest.php (hunk 2 above)
Comment (wouterj) [reply src/Symfony/Component/Security/Http/Tests/EventListener/PasswordMigratingListenerTest.php]: Thank you!

**Review by derrabus — APPROVED** [review]

**Comment by chalasr** [comment]
Thank you @wouterj.
